File: tsl_ros2_publisher_cpp/src/tsl_publisher.cpp
// Copyright 2024 Simon Sagmeister
#include "tsl_ros2_publisher_cpp/tsl_publisher.hpp"
namespace tam::tsl
{
TSLPublisher::TSLPublisher(
  rclcpp::Node * node_handle, LoggerAccessInterface::SharedPtr logger,
  const std::string & channel_suffix, CompressionStrategy::SharedPtr compression_strategy)
: node_handle_{node_handle}, logger_{logger}, compression_strategy_{compression_strategy}
{
  std::string channel_suffix_cleaned = helpers::clean_signal_name(channel_suffix);
  // Use a generic publisher since this is able to both handle ipc publish and serialized publish
  // If we would use a normal publisher we would have to disable ipc to not throw an exception
  // if the node is part of another executor.
  definition_publisher_ = node_handle->create_generic_publisher(
    std::string("/tsl") + node_handle->get_fully_qualified_name() + channel_suffix_cleaned +
      "/def",
    "tsl_msgs/msg/TSLDefinition", qos_tsl_definition_);
  value_publisher_ = node_handle->create_generic_publisher(
    std::string("/tsl") + node_handle->get_fully_qualified_name() + channel_suffix_cleaned,
    "tsl_msgs/msg/TSLValues", qos_tsl_values_);
}
void TSLPublisher::trigger()
{
  msg_creator_.process_data(logger_->get_data());

  if (def_trigger_.due()) {
    definition_publisher_->publish(msg_creator_.get_definition());
    def_trigger_.register_publishment();  // Register a published def
  }

  value_publisher_->publish(msg_creator_.get_values());
}
}  // namespace tam::tsl

File: tsl_ros2_publisher_py/tsl_ros2_publisher_py/tsl_publisher.py
# Copyright 2024 Simon Sagmeister
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from tsl_msgs.msg import TSLDefinition, TSLValues
from tsl_logger_py import LoggerAccessInterface
from tsl_ros2_publisher_py.builtin_compression_strategies import Compress64to32Bit
from tsl_ros2_publisher_py._cpp_binding import (
    MsgCreator,
    CompressionStrategy,
    clean_signal_name,
)
from tsl_ros2_publisher_py.definition_trigger import DefinitionTriggerTimer


class TSLPublisher:

    _qos_tsl_definition = QoSProfile(
        reliability=ReliabilityPolicy.BEST_EFFORT,
        history=HistoryPolicy.KEEP_LAST,
        depth=1,
    )
    _qos_tsl_values = _qos_tsl_definition

    def __init__(
        self,
        node_handle: Node,
        logger: LoggerAccessInterface,
        channel_suffix: str = "",
        compression_strategy: CompressionStrategy = Compress64to32Bit(),
    ):
        self._node_handle = node_handle
        self._logger = logger
        self._compression_strategy = compression_strategy

        self._def_trigger = DefinitionTriggerTimer(node_handle)
        self._msg_converter = MsgCreator(self._compression_strategy)

        channel_suffix_cleaned = clean_signal_name(channel_suffix)

        self._definition_publisher = node_handle.create_publisher(
            TSLDefinition,
            f"/tsl{node_handle.get_fully_qualified_name()}{channel_suffix_cleaned}/def",
            self.__class__._qos_tsl_definition,
        )
        self._value_publisher = node_handle.create_publisher(
            TSLValues,
            f"/tsl{node_handle.get_fully_qualified_name()}{channel_suffix_cleaned}",
            self.__class__._qos_tsl_values,
        )

    def trigger(self):
        self._msg_converter.process_data(self._logger)

        if self._def_trigger.due():
            self._definition_publisher.publish(self._msg_converter.get_definition())
            self._def_trigger.register_publishment()  # Register a published def

        self._value_publisher.publish(self._msg_converter.get_values())
